lib/SILOptimizer/Mandatory/DiagnoseUnreachable.cpp

```diff
@@ -746,6 +747,110 @@ static void performNoReturnFunctionProcessing(SILModule *M,
   }
 }
 
+static bool hasRecursiveCallInPath(SILBasicBlock &Block) {
+  // Process all instructions in the block to find applies that reference
+  // the parent function.  Also looks through vtables for statically dispatched
+  // (witness) methods.
+  for (auto &I : Block) {
+    auto *AI = dyn_cast<ApplyInst>(&I);
+    if (AI && AI->getCalleeFunction()
+        && AI->getCalleeFunction() == Block.getParent())
+      return true;
+
+    if (FullApplySite FAI = FullApplySite::isa(&I)) {
+      auto &M = FAI.getModule();
+      if (auto *CMI = dyn_cast<ClassMethodInst>(FAI.getCallee())) {
+        // Don't touch dynamic dispatch.
+        if (CMI->isVolatile())
+          continue;
+
+        auto ClassType = CMI->getOperand()->getType();
+        if (ClassType.is<MetatypeType>())
+          ClassType = ClassType.getMetatypeInstanceType(M);
+
+        auto *F
+            = M.lookUpFunctionInVTable(ClassType.getClassOrBoundGenericClass(),
+                                       CMI->getMember());
+
+       if (F == Block.getParent())
+         return true;
+
+        continue;
+      }
+
+      if (auto *WMI = dyn_cast<WitnessMethodInst>(FAI.getCallee())) {
+        // Don't touch dynamic dispatch.
+        if (WMI->isVolatile())
+          continue;
+
+        SILFunction *F;
+        SILWitnessTable *WT;
+
+        std::tie(F, WT)
+            = M.lookUpFunctionInWitnessTable(WMI->getConformance(),
+                                             WMI->getMember());
+        if (F == Block.getParent())
+          return true;
+
+        continue;
+      }
+    }
+  }
+  return false;
+}
+
+static bool checkInfinitelyRecursiveApplies(SILFunction &Fn,
+                     const llvm::SmallPtrSetImpl<SILBasicBlock *> &ExitBlocks) {
+  if (Fn.empty() || ExitBlocks.empty() || Fn.isThunk() != IsNotThunk)
+    return false;
+
+  enum RecursiveState {
+    FoundRecursion = 0,
+    FoundPathOutOfFunction = 1,
+  };
+
+  // Set the entry block into the state map.
+  llvm::DenseMap<SILBasicBlock *, RecursiveState> States;
+  States[Fn.getEntryBlock()] = FoundPathOutOfFunction;
+  
+  // Make the processing stack and seed it with the entry block.
+  SmallVector<SILBasicBlock *, 16> Stack;
+  Stack.push_back(Fn.getEntryBlock());
+  
+  while (!Stack.empty()) {
+    SILBasicBlock *CurBlock = Stack.pop_back_val();
```

Comment:
I wonder how this loop handles loops. Does its visit the BBs which belong to a loop multiple times? 

BTW, is it necessary to start with the entry block? Does it matter at all, which block is used as a starting point? Would using RPOT order speed-up the convergence?

Reply:
> I wonder how this loop handles loops. Does its visit the BBs which belong to a loop multiple times?

BasicBlocks can only migrate from the lower state (FoundRecursion) to higher states (FoundPathOutOfFunction) at which point they will be popped and ignored on the next turn.  If the block doesn't migrate, then it will not be added to the work queue.  Hence the analysis always converges.

For the loops I was able to come up with to convince myself of this, it wasn't visiting any of the basic blocks more than once and I wouldn't expect it to.  Even if the loop header gets re-enqueued it and its children will not be re-enqueued.